### ISSUE TITLE
JPQL native syntax for date/time literals

### ISSUE THREAD

**Issue by gavinking** [jakartaee/persistence#453]
This came up in #391.

Currently, JPQL has no native syntax for date/time literals, and just piggy-backs off the JDBC escape syntax. This syntax is kinda ugly and looks visually out-of-place in a JPQL query.

We recently added a native syntax for this to HQL, it's documented here:

<https://docs.jboss.org/hibernate/orm/6.3/querylanguage/html_single/Hibernate_Query_Language.html#datetime-literals>

I think it would be nice to add some subset of this to the spec.

Thoughts?

**Comment by trajano** [comment]
**Braced** primarily because it is shortest among them.

**Comment by gavinking** [comment]
> Braced primarily because it is shortest among them. 

It's short, but it's also the option that looks most out of place in the SQL-like syntax. I personally don't find it aesthetically appealing, though of course others are free to disagree.

> Not only that you can tell when it ends unlike the third one.

This is an advantage indeed.

**Comment by gavinking** [comment]
> Question is how do we handle `ZonedDateTime` and `Instant` but it may be implicit with `OffsetDateTime` already.

I don't think we actually need `Instant` literals, all you would need is an implicit conversion from `OffsetDateTime` to `Instant`.

That said, I *do* still think that having the type stated explicitly, as in `date 2000-12-3` is a weak advantage.

**Comment by trajano** [comment]
The closest thing to SQL is literally `'` 

That is `'2023-08-14'` and having the JPQL know that the target the type `LocalDate` and perform the necessary.  However, I still prefer `{ }` because it opens up the possibility of other things lets say 
* `UUID` 
* semantic versions {2.0.5} // I don't think any database supports that
* geo data such as {-77.00, 55.00} representing latitude and longitude
* A period `{ 2023-08-14 to 2023-08-30 }` 
* A duration `{ 16 days }`

**Comment by gavinking** [comment]
Right, this is an argument that @sebersole also makes.

**Comment by trajano** [comment]
Which one @gavinking  the use of `'` or the additional representations possible

**Comment by gavinking** [comment]
Your point about `{ .... }` being generalizable to other types.

**Comment by ajkyffin** [comment]
I naively expected single-quoted literals to work. I don't like `{ }` because that's used for arrays.

**Comment by trajano** [comment]
One other advantage of brace is it's easier to scan to make sure you didn't miss anything.  That being said why not both?  Back in Perl land we used to do

`q{foo}` to mean `'foo'`
`qq[foo]` to mean `"foo"`

Not saying we have to use the same syntax.

**Comment by gavinking** [comment]
So let me describe my thinking on this.

First, I really think the explicit type name is advantageous, for various reasons. In Hibernate we allow:

- `local time` (can be abbreviated as `time`)
- `local date` (can be abbreviated as `date`)
- `local datetime`, `offset datetime`, and `zoned datetime`  (can be abbreviated as `datetime`)

When `local time`, `local date`, or `local datetime` occur alone, they return the _current_ local time, date, or datetime, as mandated by JPA.

So, say you agree with me that using the type as a prefix is a good idea. Then we have to decide how to delimit the actual literal. Reasonable options include:

- `date '2025-5-1'`, `time '6:59:12'`
- `date (2025-5-1)`, `time (6:59:12)`
- `date {2025-5-1}`, `time {6:59:12}`
- `date [2025-5-1]`, `time [6:59:12]`

I think if you held a gun to my head and forced me to pick between these options, I would go with parentheses because they're the "default" delimiter and they look pretty nice here. I definitely think I could live with that option.

But, I have to ask, _what work are they really doing_?

Why not just:

- `date 2025-5-1`, `time 6:59:12`

That's just _way_ prettier to my eyes, and there's no ambiguity at all that could possibly trip up the parser.

Then a date time is:

- `datetime 2025-5-1 6:59`

Or, if it has an offset:

- `offset datetime 2025-5-1 6:59:12+10:00`

That's if we decide we need offset datetime literals. (We definitely don't need zoned datatime literals, IMO.)

There is one teensy tiny thing that might trip up the parser here, which is an offset with no minutes, for example:

- `datetime 2025-5-1 6:59:12+10`

This could be parsed as an integer added to a datetime literal. But since that doesn't make sense, this ambiguity is easily resolved at the semantic level, or, alternatively, we just require one of the following:

1. `offset` is mandatory for offset datetime literals, or
2. you must explicitly specify minutes in an offset.

So this really isn't anything to worry about. To be clear, I don't think offset datetime literals are common _at all_, and I'm not sure the spec needs to say anything about them.

So I'm inclined to propose that instead of us arguing over which of `''` `{}` `()` is our favorite pair of delimiters, we sidestep that debate by just not having delimiters.

**Comment by trajano** [comment]
parenthesis but allow `date(2025-02-01)` that is no space.

`date '2025-02-01'` may be okay good as an alternative too.

but it's still a bit verbose I think `d` and let it be parsed from there I think you can determine from the string what type it is e.g. `2021-02-01` would be reasonably a local date
`2021-02-01T22:00:12.231231` would reasonably be a local date time
`22:00:12.231231` would reasonably be a local time
`2021-02-01T22:00:12.231231Z` would reasonably be a ZonedDateTime

If the heuristic won't match then
```
d '2021-02-01' as datetime
```
could coerce it to 
```
d '2021-02-01T00:00:00'
```
But as far as parser goes I think just limit it to ISO formats, so don't do `2025-1-1` I think Java's parser should handle most of it already.  But the only issue is heuristics which can likely be simple regex match.

**Comment by gavinking** [comment]
> but it's still a bit verbose I think `d` and let it be parsed from there 

Oh no not this. 

We're not optimizing language design to minimize character count anymore. If you're concerned about verbosity, count _tokens_, not characters.

It's readability/understandability that we're actually supposed to be chasing. This is a hill I will absolutely die on.

**Comment by gavinking** [comment]
Look, if I put these things side-by-side:

```hql
select d '2021-02-01T00:00:00' from ...
```
```hql
select datetime 2021-02-01 00:00:00 from ... 
```

one of them hurts me eyes and the doesn't.

**Comment by trajano** [comment]
I wonder can this also expand to semver literals like

```
select ver 1.0.0.patch2 from dual
```

**Comment by trajano** [comment]
The only thing that bothers me in your example @gavinking 
> select datetime 2021-02-01 00:00:00 from ... 

was generalizing the parser so understand that datetime has two components.  Unless we limit it to ISO format which has a `T` to separate the date and time then we can say the next token represents the value specified by the literal type.

We can then have additional capabilities to provide different types with vendor specific ones being `x_` prefixed.

**Comment by gavinking** [comment]
I mean if the purpose were to add a syntax which works for arbitrary user- or implementation-defined types then of course the design constraints are going to be different. But:

1. JPQL doesn't even define that sort of extensibility and 
2. in the actual concrete cases I can think of (arrays, vectors, spatial, json, UDTs) you have things which are structured types, and which don't naturally belong inside a quoted string. For stuff like that you would want to be able to write literal lists `[1, 2, 3]` and structures `{foo:1, bar:"hello"}`.

What makes date/time values special is that:

- they actually come built-in to JPQL,
- they have a conventional syntax which is distinctive and slightly-complicated,
- they're extremely common (though date/time *literals* only really occur in tests),
- the JPQL implementation can actually to some meaningful validation of the literal value before sending it to the database (and we want that to happen, because it results in much better error messages) and the IDE can highlight them.

In the grammar of HQL we indeed have a placeholder syntax for "extensible literal values" but in fact we have not so far found a use for it. So I don't think we need that here.

**Comment by gavinking** [comment]
Uff, it just hit me that this issue is way higher-priority than I thought it was. I had forgotten that the _type_ of literals like `{d '2025-05-01'}`, `{t '12:30:15'}`, `{ts '2025-05-01T12:30'}` are `java.sql.Date`, `java.sql.Time`, and `java.sql.Timestamp`, which we've deprecated, and that as of today JPA offers no way to write a literal `LocalDate`, `LocalTime`, or `LocalDateTime`. We definitely need to fix that in JPA4! But of course we also need to coordinate with the Jakarta Query group on that.

**Comment by trajano** [comment]
One of the valuable aspects of having open technical discussions is that they often lead to new realizations or highlight overlooked issues. Collaborative dialogue tends to surface these insights naturally.

**Comment by sebersole** [comment]
> I naively expected single-quoted literals to work. I don't like `{ }` because that's used for arrays.

Use of braces like here is actually defined by JDBC also for exactly this purpose fwiw.  So there is "precedence".

**Comment by sebersole** [comment]
> I would go with parentheses because they're the "default" delimiter and they look pretty nice here. I definitely think I could live with that option.
> But, I have to ask, what work are they really doing?

 IMO we either use an escape-style syntax (braces or parens) or we use the keyword delimiter approach.  But I agree that mixing them does not make much sense.

I do like the escape syntax for the extensibility case.  Like imagine letting the user register a "parser" for the escape token, something like:

```java
new HibernatePersistenceConfiguration(...)
        .queryLiteralParser( "uuid", MyUuidParser.class )
        ...;
...
session.createSelectionQuery( "select {uuid '....'} )...
```

But ofc, that does not need to be the answer for these built-in types, especially wrt temporal-related values.  Here I'm actually more inclined to go for a keyword (aka, "explicitly typed literal syntax) approach.

**Comment by gavinking** [comment]
@sebersole 

> I do like the escape syntax for the extensibility case. Like imagine letting the user register a "parser" for the escape token, something like:
> But ofc, that does not need to be the answer for these built-in types, especially wrt temporal-related values. Here I'm actually more inclined to go for a keyword (aka, "explicitly typed literal syntax) approach.

One way of looking at this is that if we "use up" `{ }` or `[ ]` for literal dates and times, we can't use them for other things like, e.g. JSON object literals or spatial types or whatever. (Or at least, not without a lot difficulty.)

If we leave `{ }` as uninterpreted tokens with no meaning in Jakarta Query, it means that implementations can freely use these tokens for other vendor-specific stuff.

**Comment by sebersole** [comment]
Yep, I had the same thought.  Which is why I said not using them here is probably better -

> Here I'm actually more inclined to go for a keyword (aka, "explicitly typed literal syntax") approach.

aka `date yyyy-mm-dd`.  

Plus, combined with `cast` (and well-defined rules) you could already get all types (`cast` need not always be passed to the db).  E.g. `cast( datetime '2025-05-02 01:01:01' as instant )`

**Comment by gavinking** [comment]
Closing in favor of Jakarta Query issue https://github.com/jakartaee/query/issues/99.
